mgexpose/downstream.py

```diff
@@ -0,0 +1,488 @@
+    return nested
+
+
+def stat_core(islands):
+    """Calculate statistics on MGEs in the core genome and return as a dictionary."""
+    total = 0
+    core = 0
+    for island in islands:
+        if island.is_core:
+            core += 1
+        total += 1
+    core_prop = (core / total * 100) if total > 0 else 0
+    acc_prop = ((total - core) / total * 100) if total > 0 else 0
+
+    return {
+        "total": total,
+        "core_count": core,
+        "accessory_count": total - core,
+        "core_prop": core_prop,
+        "acc_prop": acc_prop,
+    }
+
+
+def count_core(islands):
+    """
+    Calculate the count of MGEs in the core genome.
+    
+    Parameters:
+    - islands: List of MGE objects.
+    
+    Returns:
+    - Integer count of MGEs in the core genome.
+    """
+    core = sum(1 for island in islands if island.is_core)
+    return core
+
+
+def count_total_islands(islands):
+    """
+    Count the total number of MGE islands.
+    
+    Parameters:
+    - islands: List of MGE objects.
+    
+    Returns:
+    - Integer count of total MGE islands.
+    """
+    return len(list(islands))
+
+
+def stat_mge_type(islands):
+    """
+    Calculate counts of each MGE type and return as a Counter object.
+    
+    Parameters:
+    - islands: List of MGE objects.
+    
+    Returns:
+    - Counter object with counts of each MGE type.
+    """
+    mge_counts = Counter()
+    for island in islands:
+        try:
+            if island.mge_type == "nested":
+                mge_counts["nested"] += 1
+            else:
+                # Get the first key (assuming there's only one key)
+                mge = next(iter(island.mge.keys()))
+                mge_counts[mge] += 1
+        except Exception as e:
+            raise ValueError(f"Unknown or absent MGE type: {e}")
+    
+    return mge_counts
+
+
+def stat_mean_genes(islands):
+    """Calculate the mean number of genes per MGE and return as a dictionary."""
+    genes_lst = [island.n_genes for island in islands]
+    mean_genes = (sum(genes_lst) / len(genes_lst)) if genes_lst else 0
+
+    # Return the result as a dictionary
+    return {"mean_genes_per_mge": mean_genes}
+
+
+def extract_cargo(island):
+    cargo_genes = []
+    for gene in island.genes:
+        if (gene.phage is None) and (gene.recombinase is None) and (gene.secretion_system is None):
+            cargo_genes.append(gene)
+    return cargo_genes
+
+
+def get_kegg_ko(gene):
+    for key, value in gene.eggnog:
+        if key == "kegg_ko":
+            return value
+
+
+def get_cazy(gene):
+    for key, value in gene.eggnog:
+        if key == "cazy":
+            return value
+
+
+def get_cog_category(gene):
+    if gene.eggnog:
+        for key, value in gene.eggnog:
+            if key == "cog_fcat":
+                if value:
+                    return value
+                else:
+                    return '-'
+            else:
+                return 'no_cog_fcat'
```

Comment:
This will always return `no_cog_fcat` if `cog_fcat` is not the first encountered record in the eggnog annotation.